cmd/kubeadm/app/phases/upgrade/prepull.go

```diff
@@ -175,18 +175,30 @@ func buildPrePullDaemonSet(component, image string) *apps.DaemonSet {
 					},
 				},
 				Spec: v1.PodSpec{
+					// Use an init container to prepull the target component image.
+					// Once this completes, a regular container with "pause" will start to get this
+					// Pod in Running state with a blocking container process. Note that DaemonSet
+					// Pods can only use RestartPolicy of Always, so there has to be a blocking
+					// process to achieve the Running state. The regular container can only start
+					// if the init container that performs the component prepull succeeded.
+					InitContainers: []v1.Container{
+						{
+							Name:  component,
+							Image: image,
```

Comment:
That works for now, but it's likely not going to work in the long term with distroless.
It works ATM, because the entry point of k8s images is `/bin/sh` with no params. When this init container is started, it's initialized without stdin. Hence, `/bin/sh` has nothing to do and exits with status code of 0 (success). The init container then terminates gracefully and the main pause container is started.

In the case of distroless there won't be any `/bin/sh` present. Hence, the entry point is likely going to be the component binary itself. Without any command line args and a component config file, this would exit with an error (likely displaying command line help). Then the init container would fail and the pods would be in an init crash loop. The `pause` container would never get executed in that case.

Hence, we would be back to square one, except we now have an init container too.

Reply:
P.S. Here's an example that simulates what I am talking about:
```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: test-ds
  namespace: kube-system
  labels:
    k8s-app: test-ds
spec:
  selector:
    matchLabels:
      name: test-ds
  template:
    metadata:
      labels:
        name: test-ds
    spec:
      tolerations:
      - key: node-role.kubernetes.io/master
        effect: NoSchedule
      initContainers:
      - name: kube-apiserver
        image: k8s.gcr.io/kube-apiserver:v1.17.0
        command: 
        - /bin/false
      containers:
      - name: pause
        image: k8s.gcr.io/pause:3.1
        command:
        - /pause
      terminationGracePeriodSeconds: 0
```

Reply:
> Without any command line args and a component config file, this would exit with an error (likely displaying command line help).

@rosti would it work if we execute `componentname --version` as the command?